### ISSUE TITLE
Missing client token with approle

### ISSUE THREAD

**Issue by madchap** [hvac/hvac#298]
Hi,

Vault 0.11.3 and hvac 0.6.4.

Probably not an issue, but this "used to work before" (like Vault 0.10 and hvac 5.x).

Reproducing the following flow: 
![approle](https://www.datocms-assets.com/2885/1517516192-vault-approle-workflow2.png)

Piece of the code failing:
```
def get_vault_token(vault_rolename):
    # get role id from vault-tower
    role_id_json = requests_wrap.get_data("{}{}/{}".format(vault_token_tower, "/roleid", vault_rolename))
    role_id = role_id_json['role_id']
    print("roleid: {}".format(role_id))

    # get secret wrap from vault-tower
    wrapped_token_json = requests_wrap.post_data("{}{}/{}".format(vault_token_tower, "/wraptoken", vault_rolename))
    wrapped_token = wrapped_token_json['wrap_token']
    print("wrapped token: {}".format(wrapped_token))

    # unwrap token to get secretid
   # Failing here
    unwrapped_secret_id = vc.unwrap(wrapped_token)
    print(unwrapped_secret_id['data']['secret_id'])
    print("Secret id: {}".format(unwrapped_secret_id))
```

with the following -- see https://gist.github.com/madchap/f7efd023c9c6f6058c4b8192aa97323c

The policy associated with the role is the following:
```
# Login with AppRole
path "auth/approle/login" {
  capabilities = [ "create", "read" ]
}

path "auth/approle/role/torrent9scraper" {
  capabilities = [ "create", "update", "read" ]
}
```

Basically, any action with object `vc` fails with that error (i.e- `vc.set_role_id()`, or `vc.create_token()`) The `vc` Client only has the `url` param passed to it, no token obviously. I am using a simulated "trusted entity" to map to the schema.

With my "admin" profile from the CLI, I am able to unwrap with no problem.

I am a bit at a loss here. I definitely used to be able to unwrap a token this way before.

Thanks for shedding some light.
Cheers,
fred

**Comment by madchap** [comment]
OK, so, I've found my issue.

Code that used to work:
```
unwrapped_secret_id = vc.unwrap(wrapped_token)
```

Code that now works:
```
vc.token = wrapped_token
unwrapped_secret_id = vc.unwrap()['data']['secret_id'])
```
Is this a bug?

**Comment by jeffwecan** [comment]
@madchap: Not sure it qualifies as a legit bug, but the behavior in question has changed semi-recently. The associated PR appears to be https://github.com/hvac/hvac/pull/149. At the very least, we should have realized the updated method would break folk's usage of it and called that out more explicitly in the changelogs.

More generally, I've been wondering whether these sort of methods should ideally be returning the full response from the Vault API or just the relevant bits of the response. I.e., just the unwrapped payload or the full response in your particular example. What would be your preference in this case?

**Comment by madchap** [comment]
> More generally, I've been wondering whether these sort of methods should ideally be returning the full response from the Vault API or just the relevant bits of the response. I.e., just the unwrapped payload or the full response in your particular example. What would be your preference in this case?

Well, I am not a very experienced programmer or anything. I am guessing the more info, the better somehow, as long as it does not give away sensitive info? However, the message in this case was quite clear. I just got stubborn in thinking "it just worked before", and I basically got around to "did I really understand how AppRole is supposed to work, or what?" (sometimes, could still have some doubts ;-))

I saw the changelog for 0.6.1, yet it didn't trigger any further thought process, as the documentation didn't change at https://hvac.readthedocs.io/en/latest/source/hvac_v1.html?highlight=unwrap (still shows `unwrap(token=None)`. (Unless I am not reading the right place).

**Comment by espoelstra** [comment]
I think providing an API similar to how the Vault CLI works would probably be the most useful for people. Perhaps following how the Requests library provides responses as an object with response.code for the status code, response.json for the json return, and response.text to dump it in a slightly nicer format for direct printing.
Maybe `hvac` could wrap responses as a Python OrderedDict by default (simple_salesforce does this against the Salesforce API) and then offer an option to just get the raw JSON as mentioned above.

**Comment by espoelstra** [comment]
I'm actually here because I'm having a similar issue with the only difference being my `secret_id` is already unwrapped due to a slightly different use case. Currently as our operations group is rolling out Vault as a POC they are issuing the `role_id` and `secret_id` both wrapped with a very short TTL with the expectation that the requester of the AppRole will unwrap them manually and provide them securely to their application. The trouble is that `hvac` doesn't seem to properly utilize the unwrapped `secret_id`, complaining about a missing token even though I know for a fact that it has a valid `role_id` and `secret_id` as I've used them to request a token manually.

I'd imagine the trouble is in the `__init__` where it is probably looking for /requiring a token before making the initial Client connection when that should be optional during the init, but required for calls other than `auth_approle` I'd imagine.

**Comment by espoelstra** [comment]
I may have gotten slightly farther by putting `token=''` into the initial constructor, now I'm fighting the AppRole policies which may not have gotten created properly. I'll see if I can set up my own dummy instance of Vault with AppRoles/Namespaces and see how hvac handles the mix. I noticed looking at the code that "use_token=True" was scattered through a lot of methods, it seems like it might be useful to create an exception for "InvalidMissingVaultToken" and simply have any methods that need an actual token to raise the exception if it wasn't defined via `auth_approle` or passed in through the environment.

**Comment by jeffwecan** [comment]
Thanks for the additional context and suggestions @espoelstra! I am aiming to replicate the scenario being described in this issue and try to work out a better flow for this type of use-case over the next couple of weeks (and ideally include whatever changes result in the upcoming v0.7.3 release).

That said, I would also welcome any sort of suggested changes in a pull request from any interested parties as well! 😉

**Comment by espoelstra** [comment]
OK, so I think I figured my issue after looking over the code and comparing it to a shell script used by our Ops folks for setting up and testing AppRoles and Namespaces.

The `mount_point` kwarg was confusing me when it might be more an "auth_url_path" as it goes in `auth/{auth_url_path}/login` in the code and I'm not sure if it also determines the local "mount" path or not?

I didn't see the namespace specifically called out in the `auth_approle` configuration (I believe it isn't required there), and the `auth_url_path` can be separate/different from the Namespace because it is related to the AppRole/PolicyName rather than the Namespace. The Namespace is used when querying the secrets to get you into the right "bucket", while the AppRole's policy limits the paths within that bucket that you can traverse. In our case the policy actually created a new "bucket" within the Namespace that was more tightly scoped to the specific Application using the AppRole rather than the entire Namespace being accessible.

I'm not currently using the `unwrap` functionality related to the current issue so take my meanderings with a grain of salt. Sadly the project I was using this for is winding down so I don't know if I'll get a chance to write more tests to validate my new assumptions.

**Comment by jeffwecan** [comment]
Thanks for the input @reiven! FWIW, along those lines, the follow code snippet is what I've personally ended up with in a Django app hosted via Heroku that receives a wrapped secret_id and then instantiates a hvac client using those credentials (this may not be directly related to the most recent comments on this issue, but I figured it would be worth chiming in with my own usage in case its helpful for any other folks while we pin down what problem(s) this issue should be solving for):


```python
import logging

import hvac
from django.conf import settings
from django.core.cache import cache
from hvac.exceptions import VaultError

from apps.vault_client.models import SecretId

logger = logging.getLogger(__name__)

VAULT_TOKEN_CACHE_KEY = 'vault_token'


def verify_then_unwrap_secret_id(response_wrapping_token):
    """Validate a response wrapping token to verify its integrity and then returns its containing secret_id.
    :param response_wrapping_token: str, Vault token that has "wrapped" some underlying data.
    :return: dict, Once the wrapped token has been verified and unwrapped, this method returns a "data" dictionary
    containing a secret_id_accessor and secret_id.
    """
    vault_client = hvac.Client(
        url=settings.VAULT_URL,
    )

    # First lookup the token to ensure it hasn't already been unwrapped (which *could* be bad).
    lookup_response = lookup_wrapped_token(response_wrapping_token=response_wrapping_token)

    # Then verify the creation path string matches expectations.
    # That may catch if some other entity attempted to rewrapped the token in transit.
    if not lookup_response['data']['creation_path'].startswith(settings.VAULT_WRAP_CREATION_PATH_PREFIX):
        error_msg = 'Token creation path "{}" does not match expected path prefix "{}". Unable to continue.'.format(
            lookup_response['data']['creation_path'],
            settings.VAULT_WRAP_CREATION_PATH_PREFIX,
        )
        raise AssertionError(error_msg)

    # Now unwrap the token and verify we received a secret_id.
    unwrapped_response = vault_client.unwrap(token=response_wrapping_token)
    assert 'secret_id' in unwrapped_response.get('data'), '"secret_id" not found in unwrapped_response data.'

    # Attempt to authenticate using the new secret_id. If successful, return the secret_id for later token generation.
    auth_response = vault_client.auth_approle(
        role_id=settings.VAULT_ROLE_ID,
        secret_id=unwrapped_response['data']['secret_id'],
        mount_point=settings.VAULT_APPROLE_MOUNT_POINT,
        use_token=False,
    )
    assert 'client_token' in auth_response.get('auth'), 'Unable to retrieve a "client_token" with provided secret ID.'

    return unwrapped_response.get('data')


def lookup_wrapped_token(response_wrapping_token):
    """Lookup a provided response_wrapping_token and return associated metadata (provided the token has not
    already been unwrapped).
    TODO: replace this with a explicit method in hvac upstream.
    :param response_wrapping_token: str, Vault token that has "wrapped" some underlying data.
    :return: dict, Returns wrapping token properties.
    """
    vault_client = hvac.Client(
        url=settings.VAULT_URL,
    )
    vault_client.token = response_wrapping_token
    token_lookup = vault_client.write(
        path='sys/wrapping/lookup',
    )
    logger.debug('lookup_wrapped_token response: {}'.format(token_lookup))
    return token_lookup


def retrieve_client_token():
    """Attempt to retrieve an existing token from the Django cache if available. 

    If a token is not found in the cache, a new token is generated using the AppRole backend.
    """
    cached_token = cache.get(VAULT_TOKEN_CACHE_KEY)
    if cached_token is None or cached_token.get('client_token') is None:
        vault_client = hvac.Client(
            url=settings.VAULT_URL,
        )
        logger.info('No Vault token found in cache (under key "{}")'.format(VAULT_TOKEN_CACHE_KEY))
        latest_secret_id = SecretId.objects.latest('created')
        auth_response = vault_client.auth_approle(
            role_id=settings.VAULT_ROLE_ID,
            secret_id=latest_secret_id.secret_id,
            mount_point=settings.VAULT_APPROLE_MOUNT_POINT,
            use_token=True,
        )
        if vault_client.is_authenticated():
            cache.set(
                key=VAULT_TOKEN_CACHE_KEY,
                value=auth_response['auth'],
                timeout=auth_response['auth']['lease_duration']
            )
        else:
            error_msg = 'unable to retrieve a new Vault client token'
            logger.error(error_msg)
            raise VaultError(error_msg)
        return auth_response['auth']['client_token']
    else:
        return cached_token['client_token']


def retrieve_authenticated_client():
    """Retrieve a current Vault token and ensure that token is able to authenticate with the
    Vault service.
    :return: hvac.Client with valid token
    """
    vault_client = hvac.Client(
        url=settings.VAULT_URL,
        token=retrieve_client_token(),
    )
    if not vault_client.is_authenticated():
        error_msg = 'unable to retrieve an authenticated Vault client'
        logger.error(error_msg)
        raise VaultError(error_msg)

    return vault_client
``` 

[I haven't actually exercised this code in some time, so I'll try to loop back on that front and verify if its still relevant for the issue at hand at some point...]

**Comment by gmaurer** [comment]
Just found this and am having the same issue trying to get past the auth_approle

`client = hvac.Client(url=VAULT_URL)`
`client_auth = client.auth_approle(role_id, secret_id)`

It throws the exception 'missing client token'  I tried @reiven solution with no success yet (That may be a permission issue on my end though) . Has there been any proposed solution for this yet?
Thanks

**Comment by aaronjolson** [comment]
I am having this issue as well. I have two easy examples. 
The first one, I make the request via http/python requests The second, I attempt to use the hvac client.  The http method works fine, I receive a 200 with the expected response. The second, I get the hvac.exceptions.InvalidRequest: missing client token exception when trying to make the  client.auth_approle method call.

```
def test_vault_api():
    postbody = {
        "role_id": vault_role,
        "secret_id": vault_secret
    }
    response = requests.post(vault_login_url, json=postbody, verify=False)
    print(f"Response is {response.status_code}")
    print(response.content)


def test_hvac():
    client = hvac.Client(
            url=vault_base_url,
            verify=False
        )
    client.auth_approle(role_id=vault_role, secret_id=vault_secret)
```

The only difference between the two urls, is that in the http variable, I have the 'auth/approle/login' part included, whereas with the base_url, I leave that part off because the auth_approle method is supposed to append that part automatically.

**Comment by rljohnsn** [comment]
Also experiencing issues with hvac client auth_approle. Python 3.7 and hvac==0.10.1

```python
# does not work, use_token default is True
resp = client.auth_approle(role_id=role, 
                           secret_id=secret, 
                           use_token=True)  
# returns false
client.is_authenticated() 

# works
resp = client.auth_approle(role_id=role, 
                           secret_id=secret,
                           use_token=False)  
 # returns true
client.is_authenticated()
```
In both scenarios resp object returns with the valid temp client_token

**Comment by fhnbg** [comment]
i have the same issues. 
Adding the mount_point parameter (path to approle) resolved it:
```
 client.auth_approle(config.get('vault', 'approle'),
                        secret_id=config.get('vault', 'secret_id'), mount_point="approletest")

```

**Comment by kaypeter87** [comment]
@briantist Is this logic issue still prevalent in the latest release?

**Comment by briantist** [comment]
It seems this issue refers to the old `client.auth_approle` method instead of the newer `client.auth.approle.login` method.

The former had been deprecated for a long time and was removed in `v1.0.0`, in this commit: https://github.com/hvac/hvac/commit/56d8216a3bad7fd58bd853cef05e878d7550f537

I'll close this out, thanks!
